### PR TITLE
Scala 3 support

### PR DESCRIPTION
**Pull request by yanns**
- [x] sbt configuration
- [x] [core] structure for compilation without any features
- [ ] [core] implementation of macros in scala 3
- [ ] [core] all tests passed on scala 2 and scala 3
- [ ] implementation of macros in scala 3
- [ ] all tests passed on scala 2 and scala 3

### CODE REVIEW

**Comment by yanns** [comment]
I have a very strange issue: when trying to use `inline` on a scala file in `scala-3`, scalac does not accept it:
```
sbt:parboiledCore> scalaVersion
[info] 3.0.1
sbt:parboiledCore> compile
[error] .../parboiled2/parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:44: error: illegal start of definition
[error]   inline val pi = 3.141592653589793
[error]   ^: /Users/yannsimon/projects/parboiled2/parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala
```
I've used a similar approach in different projects without this issue. Strange.
I've tried to cleanup, but it does not help.

Maybe the cross plugin is interfering with scalac here?

**Comment by yanns** [comment]
When I switch off the auto-formatting, it's working.
https://github.com/sirthias/parboiled2/pull/274/commits/62027badadf5561329f73e5583e81243abca14b0

**Comment by yanns** [comment]
The first very very simple test is passing: https://github.com/yanns/parboiled2/commit/afc53f653792d464e0701e64f3e17746e32749f2
One macro is still missing as I have issues with dependent return type: https://discord.com/channels/632150470000902164/875868146949554207/878304800184340540

**Comment by lrytz** [comment]
Minimization:
```scala
import scala.quoted.*
trait C:
  type T
  def foo: T
inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
def impl[CC <: C](xp: Expr[CC])(using Quotes): Expr[CC#T] = '{ $xp.foo }
```

```
-- Error: Test.scala:5:45 ------------------------------------------------------
5 |inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
  |                                             ^
  |                           access to parameter x from wrong staging level:
  |                            - the definition is at level 0,
  |                            - but the access is at level -1.
-- Error: Test.scala:5:53 ------------------------------------------------------
5 |inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
  |                                                     ^
  |                           access to parameter x from wrong staging level:
  |                            - the definition is at level 0,
  |                            - but the access is at level -1.
```

(only the first error shows if the `[x.type]` argument left to inference)

@nicolasstucki do you have an idea how this could be expressed?

**Comment by nicolasstucki** [comment]
> Minimization:
> 
> ```scala
> import scala.quoted.*
> trait C:
>   type T
>   def foo: T
> inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
> def impl[CC <: C](xp: Expr[CC])(using Quotes): Expr[CC#T] = '{ $xp.foo }
> ```
> 
> ```
> -- Error: Test.scala:5:45 ------------------------------------------------------
> 5 |inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
>   |                                             ^
>   |                           access to parameter x from wrong staging level:
>   |                            - the definition is at level 0,
>   |                            - but the access is at level -1.
> -- Error: Test.scala:5:53 ------------------------------------------------------
> 5 |inline def makro(inline x: C): x.T = ${ impl[x.type]('x) }
>   |                                                     ^
>   |                           access to parameter x from wrong staging level:
>   |                            - the definition is at level 0,
>   |                            - but the access is at level -1.
> ```
> 
> (only the first error shows if the `[x.type]` argument left to inference)
> 
> @nicolasstucki do you have an idea how this could be expressed?

Seems to be a variant of https://github.com/lampepfl/dotty/issues/8100, the current wip fix in https://github.com/lampepfl/dotty/pull/12540 does not fix this variant.

I opened https://github.com/lampepfl/dotty/issues/13376 to track this issue.

I could not find a workaround.

**Comment by lrytz** [comment]
Thanks @nicolasstucki.

To unblock things, this workaround could be used for now:

```scala
    inline def run()(implicit scheme: Parser.DeliveryScheme[L]): scheme.Result = ${
      ParserMacros.runImpl[scheme.Result]()('scheme)
    }
```

```scala
  def runImpl[R]()(schemeExpr: Expr[Parser.DeliveryScheme[_]])(using Quotes): Expr[R] =
    '{
      $schemeExpr.failure(new Exception("TODO"))
    }.asInstanceOf[Expr[R]]
```

The casting is only in private code, the public `run` method has the desired signature.

**Comment by yanns** [comment]
In scala 2, we can write macro like this:
```
val unwrappedTree = q"cursorChar == $charTree && __advance()"
```

The variable `cursorChar` and the function `__advance` are calls on the `$this`  instance of a `Parser`.

In scala 3, the quotes are checked. Using `${ __advance() } ` directly cannot compile as the macro does not know that we are "inside" an instance of a `Parser`.
To fix that, I extracted the parser instance:
```
    val opTree: OpTree = r match {
      case '{ ($p: Parser).ch($c) } =>
        CharMatch(p, c)
      case _ => reportError("does not understand", name)
    }
```
so that I can write quotes like this:
```'{  $parser.cursorChar == $charTree && $parser.__advance() }```

I'm wondering if using `$parser.__advance()` instead of `__advance()` can have some bad implications, maybe on performances. If anyone has an opinion about it, it'd be the right time to express it as we will add more & more quotes using this pattern.

**Comment by nicolasstucki** [comment]
> https://github.com/sirthias/parboiled2/pull/274#issuecomment-905261279

Should not use a cast on the `Expr`

```scala
  '{
      $schemeExpr.failure(new Exception("TODO"))
    }.asInstanceOf[Expr[R]]
      ^^^^^^^^^^^^
```

It is better to add the cast in the generated code

```scala
  '{ $schemeExpr.failure(new Exception("TODO")).asInstanceOf[R] }
```

Or alternatively 
```scala
  '{ $schemeExpr.failure(new Exception("TODO")) }.asExprOf[R]
```
Not sure if this last one would work in this case

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  case class Sequence(ops: Seq[OpTree]) extends OpTree {
+    val parser = ops.head.parser
+    override def render(wrapped: Boolean)(using Quotes): Expr[Boolean] =
+      ops
+        .map(_.render(wrapped))
+        .reduceLeft((l, r) => '{ val ll = $l; if (ll) $r else false })
+  }
+
+  sealed abstract class TerminalOpTree(parser: Expr[Parser]) extends OpTree {
```
Comment (lrytz) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:100]:
```suggestion
  sealed abstract class TerminalOpTree extends OpTree {
```

(the `parser` method is inherited from `OpTree`)

**Comment by lrytz** [comment]
> I'm wondering if using `$parser.__advance()` instead of `__advance()` can have some bad implications

It looks good, in the generated code it's `this.__advance()`.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala (hunk 1 above)
Comment (yanns) [reply parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala]: 👍 
https://github.com/sirthias/parboiled2/pull/274/commits/86a585ae47d1b45b2f213a43692a0d9da9b7a6e2

**Comment by jrudolph** [comment]
@yanns are you fine with creating a topic branch for Scala 3 support, so it's easier to work together? The idea would be to PR against this branch until it's working and then merge it into master.

**Comment by yanns** [comment]
> @yanns are you fine with creating a topic branch for Scala 3 support, so it's easier to work together? The idea would be to PR against this branch until it's working and then merge it into master.

Yes of course!
This is the goal of this branch that I've opened as a WIP PR: https://github.com/sirthias/parboiled2/pull/274
Maybe it's better to have a branch living in https://github.com/sirthias/parboiled2 ?

**Comment by jrudolph** [comment]
>  Maybe it's better to have a branch living in [sirthias/parboiled2](https://github.com/sirthias/parboiled2) ?

Yes, that's what I will do now. I will just take the current state of this branch as the starting point.

**Comment by jrudolph** [comment]
Created #280. From now we can PR/review against that branch.

**Comment by yanns** [comment]
Please continue discussions on https://github.com/sirthias/parboiled2/pull/280

**Pull request closed without merging**